### DISCUSSION TITLE
[Installation] .so .dll file not found

### DISCUSSION THREAD

**Discussion by adykstra** [jonescompneurolab/hnn-core#835] [Installation] [answered]
### Describe the problem to us.

I'm on a Macbook Pro with the M3 Pro chip running Sonoma 14.5.

I installed successfully, tested the gui (which opened in a browser - is that expected behavior?), but when I tried run the first example script (plot_simulate_evoked.py) from within ipython, it gave me this traceback:
[Dykstra_traceback_1.pdf](https://github.com/user-attachments/files/16380248/Dykstra_traceback_1.pdf)

I thought it was a joblib lissue, so I installed that using pip in the hnn environment, and got this traceback:
[Dykstra_traceback_2.pdf](https://github.com/user-attachments/files/16380251/Dykstra_traceback_2.pdf)

### What steps did you follow to install?

`conda create --name hnn`
`conda activate hnn`

From within the newly created hnn environment:
`conda install anaconda::pip`
`pip install hnn_core'[gui]'`
`ipython --matplotlib`

From the iPython shell:
`%run plot_simulate_evoked.py`

This plotted both the network and a multicompartment PN, but stopped there with the error below.

![Figure_1](https://github.com/user-attachments/assets/ad1199f0-6eff-4f53-a517-105b38ddac3c)
![Figure_2](https://github.com/user-attachments/assets/2e5b697b-5b6d-4dea-a4c8-41b58834f3a7)

I then installed joblib successfully in the hnn environment:
`pip install joblib`
Successfully installed joblib-1.4.2

### Error message

```shell
See the traceback pdfs above.
```

### OS

Mac

**Comment by rythorpe** [comment]
Hi @adykstra, this seems like it might be an issue with the NEURON mod files not getting built or getting put in a weird place. I also feel like this came up recently - something about NEURON not supporting the latest version of Python.... @ntolley @gtdang does this ring any bells, or am I making this up?

**Reply by gtdang** [reply]
The Python 3.12 issue we corrected for the upcoming release was unrelated to Neuron. Neuron has wheels for py3.12 and MacOS 11+ so it seems like they support 3.12. Seems like something funny is happening in the build process with Neuron. Not sure if it's an M3 issue. 

@adykstra Can you check the `/opt/anaconda3/envs/hnn/lib/python3.12/site-packages/hnn_core/mod` directory? Are there files in there? 

Could you run the command in your terminal: 
`find /opt/anaconda3/envs/hnn/lib/python3.12/site-packages/hnn_core/mod -name "*.so"`

It should list a file nested in the mod directory: `mod/arm64/.libs/libnrnmech.so`
This is the compiled Neuron C++ program that hnn is looking for.

**Reply by adykstra** [reply]
So there seem to be files in that directory:
<img width="762" alt="image" src="https://github.com/user-attachments/assets/afe12bac-0a9b-4774-872f-8c79ea4829c2">

But no nested .so files:
<img width="895" alt="image" src="https://github.com/user-attachments/assets/0e22505f-a3ba-4dee-b728-aef879698944">

**Reply by jasmainak** [reply]
This seems reminiscent of https://github.com/jonescompneurolab/hnn-core/issues/689

Just to verify, do you have `nrnivmodl` in your path? If you go to the above folder and type:

```sh
nrnivmodl
```

does it produce any new files in the folder or subfolders?

**Reply by adykstra** [reply]
"Command not found"

**Reply by jasmainak** [reply]
`pip install NEURON` and try again?

**Reply by adykstra** [reply]
From within base or the hnn environment?

**Reply by jasmainak** [reply]
everything should be done in the hnn environment where you attempted to install

**Reply by adykstra** [reply]
<img width="895" alt="image" src="https://github.com/user-attachments/assets/b13dc080-b7ef-44de-a3c9-2a4b0fea41cf">

**Reply by jasmainak** [reply]
and `nrnivmodl` does not work in `hnn` environment?

**Reply by adykstra** [reply]
Before I tried `pip install NEURON` it was "command not found" before but just now I got this and was asked via a gui pop-up to install command line developer tools:
<img width="1501" alt="image" src="https://github.com/user-attachments/assets/3dd36282-3e0b-4ce0-9669-8ac207166a10">

**Reply by jasmainak** [reply]
Looks like you do need it: https://www.neuron.yale.edu/neuron/download

I feel like I must have installed it at some point ... or maybe it came bundled with my mac.

**Reply by gtdang** [reply]
> Looks like you do need it: https://www.neuron.yale.edu/neuron/download
> 
> I feel like I must have installed it at some point ... or maybe it came bundled with my mac.

Homebrew will usually install it during install. I guess xtools CLT might need to be an Mac dependency with the current build... Unless we distribute the build mod though wheels as you suggested?

**Comment by jasmainak** [comment]
Our installer should throw an error if `nrnivmodl` is not found @gtdang @ntolley 

also we should consider distributing build mod files through wheels ...

**Comment by adykstra** [comment, marked as answer]
Update: 
- At first, trying `nrnivmodl` from within the hnn env returned "command not found".
- Not sure what I did in the interim, but at some point, trying `nrnivmodl` again prompted GUI-based installation of xtools developer tools
- After successful installation of the developer tools, I ran `nrnivmodl` successfully.
- i then ran `plot_simulate_evoked.py` successfully using joblib. Here's that output:

In [8]: %run plot_simulate_evoked.py
Warning: no DISPLAY environment variable.
--No graphics will be displayed.
Joblib will run 2 trial(s) in parallel by distributing trials over 2 jobs.
Warning: no DISPLAY environment variable.
--No graphics will be displayed.
Warning: no DISPLAY environment variable.
--No graphics will be displayed.
Loading custom mechanism files from /opt/anaconda3/envs/hnn/lib/python3.12/site-packages/hnn_core/mod/arm64/.libs/libnrnmech.so
Loading custom mechanism files from /opt/anaconda3/envs/hnn/lib/python3.12/site-packages/hnn_core/mod/arm64/.libs/libnrnmech.so
Building the NEURON model
Building the NEURON model
[Done]
Trial 1: 0.03 ms...
[Done]
Trial 2: 0.03 ms...
Trial 1: 10.0 ms...
Trial 2: 10.0 ms...
Trial 1: 20.0 ms...
Trial 2: 20.0 ms...
Trial 1: 30.0 ms...
Trial 2: 30.0 ms...
Trial 1: 40.0 ms...
Trial 2: 40.0 ms...
Trial 1: 50.0 ms...
Trial 2: 50.0 ms...
Trial 1: 60.0 ms...
Trial 2: 60.0 ms...
Trial 1: 70.0 ms...
Trial 2: 70.0 ms...
Trial 1: 80.0 ms...
Trial 2: 80.0 ms...
Trial 1: 90.0 ms...
Trial 2: 90.0 ms...
Trial 1: 100.0 ms...
Trial 2: 100.0 ms...
Trial 1: 110.0 ms...
Trial 2: 110.0 ms...
Trial 1: 120.0 ms...
Trial 2: 120.0 ms...
Trial 1: 130.0 ms...
Trial 2: 130.0 ms...
Trial 1: 140.0 ms...
Trial 2: 140.0 ms...
Trial 1: 150.0 ms...
Trial 2: 150.0 ms...
Trial 1: 160.0 ms...
Trial 2: 160.0 ms...
{'mu': 63.53, 'sigma': 3.85, 'numspikes': 1}
Joblib will run 1 trial(s) in parallel by distributing trials over 1 jobs.
Loading custom mechanism files from /opt/anaconda3/envs/hnn/lib/python3.12/site-packages/hnn_core/mod/arm64/.libs/libnrnmech.so
Building the NEURON model
[Done]
Trial 1: 0.03 ms...
Trial 1: 10.0 ms...
Trial 1: 20.0 ms...
Trial 1: 30.0 ms...
Trial 1: 40.0 ms...
Trial 1: 50.0 ms...
Trial 1: 60.0 ms...
Trial 1: 70.0 ms...
Trial 1: 80.0 ms...
Trial 1: 90.0 ms...
Trial 1: 100.0 ms...
Trial 1: 110.0 ms...
Trial 1: 120.0 ms...
Trial 1: 130.0 ms...
Trial 1: 140.0 ms...
Trial 1: 150.0 ms...
Trial 1: 160.0 ms...

**Reply by adykstra** [reply]
Thanks for all your help!
